✨ Feature : 회원가입 기능

buhee1029 opened the pull request:
## 💻 작업 내역 
### ✓ 회원가입 DTO 유효성 검증 25c89bf935e1b6f2a03c550a05a5fc3fb3b5c7ae

- **제약조건**
   - 빈 값 확인, 이메일 형식 확인, 비밀번호 형식 확인
   - 다른 개인정보와 유사한 비밀번호는 사용할 수 없습니다.
   - 비밀번호는 최소 10자 이상이어야 합니다.
   - 문자, 숫자, 특수문자 중 2가지 이상 포함해야 합니다.
-  **Custom Valid Annotation `@PasswordSimilar`**
   -  비밀번호가 다른 개인정보와 유사한지 확인하기 위한 커스텀 어노테이션입니다.
   - 계정이나 이메일이 비밀번호에 포함되어 있다면 예외 처리하였습니다.
   - 추후 단순히 포함이 아닌 두 문자열 간 유사도 검사 알고리즘을 적용시키는 것도 좋을 것 같습니다.
- **`@Validated` 이용한 유효성 검증**
   -  `@NotBlank` → `@Pattern` → `@PasswordSimilar` 순서로 실행되기 위하여 그룹시퀀스를 사용했습니다.
   - 그룹 시퀀스를 사용하기 위해서는 스프링에서 제공되는 `@Validated` 를 적용해야합니다.


### ✓ 회원가입 기능 31657f81c58aca4a23180b30abf90f4184ebb2f3
아래와 같은 순서로 회원가입 기능을 추가하였습니다.

1. 계정(username) 중복 확인
2. 비밀번호를 암호화하여 사용자 정보를 저장
3. 6자리 랜덤 코드 생성
4. 계정에 생성된 인증코드 저장
5. 계정과 인증코드가 담긴 응답 반환


## 🔎 PR 특이 사항

### 커스텀 예외 WantedException 수정 c1f92f63b4663295bed88085127cf5ac9ca3da1c

buhee1029 committed:
Feature: 회원가입 DTO 유효성 검증

- @Validated 이용한 유효성 검증
- Custom Valid Annotation
- 유효성 검증 그룹화 및 순서 지정

buhee1029 committed:
Test: 회원가입 DTO 유효성 검증 테스트

buhee1029 committed:
Refactor: 간단한 코드 수정

- User entity `@GeneratedValue(strategy = GenerationType.IDENTITY)` 추가
- 커스텀 예외 WantedException 생성자 추가

buhee1029 committed:
Feature: 암호화 적용하기 위한 Spring Security 설정

buhee1029 committed:
Feature: 회원가입 기능 추가

- AuthCode 도메인 생성
- 비밀번호 암호화 후 정보 저장
- 랜덤코드 생성 기능

buhee1029 committed:
Test: 회원가입 Service 테스트

kmss6905 commented on src/main/java/com/wanted/feed/user/service/JoinService.java:
```diff
@@ -0,0 +1,77 @@
+                                        .build());
+    }
+
+    private String createCode() {
```
`createCode()` 의 경우 테스트 할 수 있게 외부에 따로 두고 주입받아 사용한다면 좋을 것 같아요! 😄

hsjkdss228 commented on src/main/java/com/wanted/feed/exception/WantedException.java:
```diff
@@ -1,10 +1,12 @@
-@AllArgsConstructor
 public class WantedException extends RuntimeException{
-    private final ErrorType errorType = ErrorType.of(this.getClass());
+    private final ErrorType errorType;
+
+    public WantedException(ErrorType errorType) {
+        this.errorType = errorType;
+    }
```
10/27 회의 협의 내용

- WantedException에는 별도의 생성자를 두지 않음
- WantedException을 상속받는 사용자 정의 Exception 클래스를 정의
- [ErrorType]에 정의한 사용자 정의 Exception을 포함하는 enum 요소를 추가

[ErrorType]: https://github.com/wanted-pre-onboarding-backend-team-s/social-media-integrated-feed-service/pull/3/files#r1374008046

hsjkdss228 commented on src/main/java/com/wanted/feed/exception/ErrorType.java:
```diff
@@ -9,7 +9,10 @@
 @AllArgsConstructor
 public enum ErrorType {
-    U001("U001", "에러 메시지를 담습니다.", WantedException.class, HttpStatus.NOT_FOUND);
+    U001("U001", "에러 메시지를 담습니다.", WantedException.class, HttpStatus.NOT_FOUND),
+
+    CONFLICT_USERNAME("409", "중복된 계정입니다.", WantedException.class, HttpStatus.CONFLICT),
+    INTERNAL_SERVER_ERROR("500", "정의되지 않은 에러가 발생했습니다.", WantedException.class, HttpStatus.INTERNAL_SERVER_ERROR);
```
10/27 회의 협의 내용

WantedException을 상속받는 [사용자 정의 Exception]을 정의하는 경우, 여기 있는 ErrorType의 요소에 사용자 정의 Exception 클래스를 인자로 전달하는 요소를 추가

```java
F001("F001", "존재하지 않는 피드입니다.", FeedNotFound.class, HttpStatus.NOT_FOUND)
```

[사용자 정의 Exception]: https://github.com/wanted-pre-onboarding-backend-team-s/social-media-integrated-feed-service/pull/3/files#r1374002591

min050410 commented on src/main/java/com/wanted/feed/user/service/JoinService.java:
```diff
@@ -0,0 +1,77 @@
+    public JoinResponseDto join(JoinRequestDto joinRequest) {
+        validateDuplicateUsername(joinRequest.getUsername());
+
+        User user = userRepository.save(User.builder()
+                                            .username(joinRequest.getUsername())
+                                            .email(joinRequest.getEmail())
+                                            .password(passwordEncoder.encode(joinRequest.getPassword()))
+                                            .build());
```
joinRequest에서 toEntity 메서드로 분리하는건 어떠신지요?
서비스의 코드를 줄이고 좀더 깔끔하게 작성할 수 있을 것 같습니다!

buhee1029 replied:
인코드 비밀번호 필드 때문에 서비스단에서 작성했는데 필드를 넘겨서 dto 단에서 작성하면 깔끔할 것 같네요!!

kmss6905 commented on src/main/java/com/wanted/feed/user/domain/AuthCode.java:
```diff
@@ -0,0 +1,37 @@
+    @Builder
+    public AuthCode(String code, String username) {
+        this.code = code;
+        this.username = username;
```
code 와 username 값이 null 이 올 경우 검증은 하지 않아도 될까요 ❓❓❓

buhee1029 replied:
ApprovalRequestDto 에서 Validation 어노테이션을 사용하여 값이 들어올때 빈 값에 대한 유효성 검증하였는데 안전하게 엔티티에서도 한번 더 검증을 해도 좋을 것 같네요!

hsjkdss228 commented on src/main/java/com/wanted/feed/exception/WantedException.java:
```diff
@@ -1,10 +1,12 @@
+    public WantedException(ErrorType errorType) {
+        this.errorType = errorType;
+    }
 }
```
EOF(마지막 빈 줄)를 명시적으로 추가해 주시면 이 마이너스 표시를 안 볼 수 있어서 좋고(?), 특정 도구에서 스크립트의 마지막 줄임을 인지하는 데 도움을 줄 수 있다고 합니다.

min050410 commented on src/main/java/com/wanted/feed/user/service/JoinService.java:
```diff
@@ -0,0 +1,77 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+public class JoinService {
+
+    private final UserRepository userRepository;
+    private final AuthCodeRepository authCodeRepository;
+    private final PasswordEncoder passwordEncoder;
+
+    public JoinService(UserRepository userRepository,
+            AuthCodeRepository authCodeRepository, PasswordEncoder passwordEncoder) {
+        this.userRepository = userRepository;
+        this.authCodeRepository = authCodeRepository;
+        this.passwordEncoder = passwordEncoder;
+    }
```
`@RequiredArgsConstructor` 어노테이션으로 생성자 부분 줄일 수 있을 것 같습니다!

hsjkdss228 replied:
저같은 경우에는 의존성 주입을 드러내기 위해 코드에서처럼 생성자를 드러내는 편이었는데, 어노테이션으로 축약하면 메서드 내용에 더 집중할 수 있겠군요.

buhee1029 replied:
저도 명시하는 게 더 좋지 않을까 생각하였는데 축약하면 코드도 깔끔해지고 인우님 말대로 메서드에 집중할 수 있겠네요!!

min050410 commented on src/main/java/com/wanted/feed/user/service/JoinService.java:
```diff
@@ -0,0 +1,77 @@
+        int length = 6;
+
+        try {
+            Random random = SecureRandom.getInstanceStrong();
+            StringBuilder builder = new StringBuilder();
+
+            for (int i = 0; i < length; i++) {
+                builder.append(random.nextInt(10));
+            }
+
+            return builder.toString();
+        } catch (NoSuchAlgorithmException e) {
+            throw new WantedException(ErrorType.INTERNAL_SERVER_ERROR);
+        }
```
stringBuilder로 코드 생성하는 부분 굉장히 깔끔하네요 😃

hsjkdss228 commented on src/main/java/com/wanted/feed/exception/WantedException.java:
```diff
@@ -1,10 +1,12 @@
+    public WantedException(ErrorType errorType) {
+        this.errorType = errorType;
+    }
 }
```
EOF(마지막 빈 줄)를 명시적으로 추가해 주시면 이 마이너스 표시를 안 볼 수 있어서 좋고(?), 특정 도구를 사용하는 경우에 스크립트의 마지막 줄임을 인지하는 데 도움을 줄 수 있다고도 합니다.

buhee1029 replied:
오 그렇군요!! 마지막 빈 줄 추가하는 것을 저희 팀에 코드 컨벤션으로 추가해도 좋을 것 같네요.

min050410 commented on src/main/java/com/wanted/feed/user/validation/PasswordSimilarValidator.java:
```diff
@@ -0,0 +1,32 @@
+        String emailField = (String) new BeanWrapperImpl(value).getPropertyValue(email);
+        String passwordField = (String) new BeanWrapperImpl(value).getPropertyValue(password);
+
+        emailField = emailField.substring(0, emailField.indexOf("@"));
```
이런 것 "@" 매직넘버를 제거하면 좋다고 배웠습니다!

buhee1029 replied:
앗 맞네요! 언급해주셔서 감사합니다 😊

min050410 reviewed (commented):
개인적으로 깔끔한 코드여서 리뷰할게 별로 없네요 ㅎㅎ
저는 점심먹으러 가보겠습니당

hsjkdss228 commented on src/main/java/com/wanted/feed/user/dto/JoinRequestDto.java:
```diff
@@ -0,0 +1,43 @@
+        groups = ValidationGroups.SimilarCheckGroup.class)
+public class JoinRequestDto {
+
+    @NotBlank(message = "필수입니다.", groups = ValidationGroups.NotEmptyGroup.class)
```
`message`에 어떤 입력 필드에서 발생한 예외인지 알려주면 좋을 것 같습니다.

buhee1029 replied:
아 선화님과 협의한 내용이 있습니다.

선화님이 예외를 처리하는 메시지를 해당 필드를 가져와 [field] 은(는) [message] 로 반환되도록 코드를 작성해서 저는 그에 맞게 '필수입니다.', '형식에 맞게 입력해주세요' 라는 메시지로만 작성하기로 하였습니다.

hsjkdss228 replied:
아하 확인했습니다!

buhee1029 committed:
Refactor: 리뷰 받은 코드 수정

- @ 매직넘버로 변경
- 생성자 줄이기 위한 @RequiredArgsConstructor 추가
- dto 단에 toEntity 메서드로 분리

buhee1029 committed:
Refactor: 예외처리 코드 수정

- WantedException을 상속받는 사용자 정의 Exception 추가
- ErrorType에 정의한 사용자 정의 Exception을 포함하는 enum 요소를 추가
